packages/react-ui-core/src/Gmap/GmapBase.js

```diff
@@ -1,32 +1,107 @@
-import React, { PureComponent } from 'react'
+import React, { Component } from 'react'
 import PropTypes from 'prop-types'
-import { withScriptjs, withGoogleMap, GoogleMap } from 'react-google-maps'
+import themed from 'react-themed'
+import classnames from 'classnames'
+import startsWith from 'lodash/startsWith'
+import pickBy from 'lodash/pickBy'
+import { GmapContext } from './GmapContext'
+import { GmapInteraction } from './GmapInteraction'
 
 const DEFAULT_ZOOM = 8
 const DEFAULT_CENTER = { // Atlanta, GA
   lat: 33.7490,
   lng: -84.3880,
 }
 
-export class BaseGoogleMap extends PureComponent {
+@themed(/^Gmap/, { pure: true })
+
+export default class GmapBase extends Component {
   static propTypes = {
+    theme: PropTypes.object,
+    className: PropTypes.string,
     children: PropTypes.node,
+    center: PropTypes.shape({
+      lat: PropTypes.number,
+      lng: PropTypes.number,
+    }),
+    zoom: PropTypes.number,
+    clickableIcons: PropTypes.bool,
+    mapControls: PropTypes.shape({
+      fullscreenControl: PropTypes.bool,
+      mapTypeControl: PropTypes.bool,
+      zoomControl: PropTypes.bool,
+      streetViewControl: PropTypes.bool,
+    }),
+  }
+
+  static defaultProps = {
+    theme: {},
+    zoom: DEFAULT_ZOOM,
+    clickableIcons: false,
+    mapControls: {
+      fullscreenControl: false,
+      streetViewControl: false,
+    },
+  }
+
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      map: null,
+    }
+    this.googleMap = React.createRef()
+    this.initMap = this.initMap.bind(this)
+  }
+
+  componentDidMount() {
+    this.initMap()
+  }
+
+  initMap() {
+    const {
+      zoom,
+      mapControls,
+      center,
+      clickableIcons,
+    } = this.props
+
+    const map = new window.google.maps.Map(this.googleMap.current, {
+      zoom,
+      clickableIcons,
+      center: center || DEFAULT_CENTER,
+      ...mapControls,
+    })
+
+    this.setState({ map })
+    GmapInteraction.registerMap(map)
   }
 
   render() {
+    const { map } = this.state
     const {
+      theme,
+      className,
       children,
+      ...rest
     } = this.props
 
+    const attributes = pickBy(rest, (_, key) => startsWith(key, 'data'))
+
     return (
-      <GoogleMap
-        defaultZoom={DEFAULT_ZOOM}
-        defaultCenter={DEFAULT_CENTER}
+      <div
+        ref={this.googleMap}
+        className={classnames(
+          theme.Gmap,
+          className,
+        )}
+        data-tid="google-map"
+        {...attributes}
       >
-        {children}
-      </GoogleMap>
+        <GmapContext.Provider value={{ map }}>
```

Comment:
For my edification, can you explain why we are adding `GmapContext`? What does it buy us?

Reply:
I did read the [docs](https://reactjs.org/docs/context.html), BTW, and they say "Don’t use context just to avoid passing props a few levels down. Stick to cases where the same data needs to be accessed in many components at multiple levels."

I take it the map components can get nested deep enough to warrant using context?

Reply:
> the same data needs to be accessed in many components at multiple levels. 

This is the key part right here.  Right now we may end up with a bunch of components that all need access to to `map`.  It also makes it so we don't have to clone every single child that the `GmapBase` will be a parent to.